### PR TITLE
add prometheus monitoring foundation

### PR DESCRIPTION
**Pull request by Sing-Li**
Add support for prometheus monitoring  ( https://prometheus.io/ )

This can become our own foundation to instrument everything.  A first step in realizing #5730

Only one "sample metrics" has been added - accumulated number of messages sent 
But more can be readily added

Prometheus exposes default useful nodejs metrics 

There are excellent add on modules that can expose GC metrics, docker metrics , as well as OS metrics.

Used the following prometheus.yml for testing and development:

```
global:
  scrape_interval:     15s # By default, scrape targets every 15 seconds.

  # Attach these labels to any time series or alerts when communicating with
  # external systems (federation, remote storage, Alertmanager).
  external_labels:
    monitor: 'codelab-monitor'

# A scrape configuration containing exactly one endpoint to scrape:
# Here it's Prometheus itself.
scrape_configs:
  # The job name is added as a label `job=<job_name>` to any timeseries scraped from this config.
  - job_name: 'prometheus'


    # Override the global default and scrape targets from this job every 5 seconds.
    scrape_interval: 5s
    metrics_path: /api/metrics
    static_configs:
      - targets: ['127.0.0.1:3000']
```

### CODE REVIEW

**Comment by rodrigok** [comment]
Should we replace all datadog metrics by prometheus?

**Comment by geekgonecrazy** [comment]
@rodrigok I cast my vote 100% for replacing datadog with prometheus.  I think this will let us fine tune what we want / need to monitor within rocket.chat its self too.

**Comment by Sing-Li** [comment]
one gotcha to share when hooking up prometheus is that for multi-instanced install - you need to connect direct to instance [not through load balancer] to get meaningful metrics feed

**Comment by graywolf336** [comment]
If we're going to add more stats this way, then let's move them all to a new `/api/prometheus/` style api and not keep adding things to the default api.

**Comment by geekgonecrazy** [comment]
Should we go so specific as to `/api/prometheus/` or can we not just use: `/api/metrics/` ?

**Comment by graywolf336** [comment]
How generic is the data returned?

**Comment by geekgonecrazy** [comment]
@graywolf336 I think your right.  Very least should move to: `/api/metrics/prometheus`

**Comment by Sing-Li** [comment]
Not quite an "api" ... it is a data-feed that is accessed via http / https 
As new metrics are added they will become part of the feed, so no extra APIs will be added.
Format is pretty standard monitoring data feed (counter, gauge, histogram, summary).
Default is actually `/metrics`  for prometheus.  (https://prometheus.io/)

**Comment by graywolf336** [comment]
While I can understand what it is you're trying to do, it should be implemented in a manner which is extensible and easily supports additional metric systems beyond the ones that support the prometheus format. Before this makes it into production, I would like to see the changes this pull request added changed to use internal the `RocketChat.callbacks` which are not only `defer`ed but also make sense in an architectural standpoint to collect metrics from. Also, the endpoint which is used I would like to see use the structure @geekgonecrazy recommended above which is `/api/metrics/{service}`.

cc: @engelgabriel

**Comment by engelgabriel** [comment]
@Sing-Li what other system use the same data feed format as Prometheus?

**Comment by engelgabriel** [comment]
I think that the API should be a wrapper around our own stats collector, and not yet another stats collecting engine. This way we can support multiple system, and even show some basic data and graphs on the admin panel.

**Comment by graywolf336** [comment]
@engelgabriel I agree :+1:

**Comment by Sing-Li** [comment]
@graywolf336  As stated previously,  as a single endpoint datafeed - you are free to locate it anywhere.  It is not an API.

@engelgabriel  it actually is.  The client part of prometheus actually does not care how metrics are collected.  It is just a data structure (a buffer) that is updated and then exported/rendered to the feed.  How you collect those metrics, is absolutely flexible and not dictated in any way - so if @graywolf336  chooses to use `defer` or `callback`, he certainly can.

Prometheus solves the instrumentation and monitoring problem elegantly, and allows for minimal disturbance to existing systems for its implementation - hence its growing popularity.

**Comment by engelgabriel** [comment]
I'd like to have something like https://github.com/RocketChat/Rocket.Chat/issues/3824 running internally, powered by our own https://github.com/RocketChat/Rocket.Chat/issues/726

**Comment by Sing-Li** [comment]
@engelgabriel   as most of the metrics enumerated in #726 are counters, and the https://moovel.github.io/teamchatviz/ site referenced in #3824 are time series graphs ... this is almost a classic use-case for a prometheus - Grafana pipeline.

But now I do understand (thx 4 the offline talk) that there are some plans already in motion to accomplish some part of this using REST APIs and other means.

So indeed it is your call on the approach - to avoid duplication of effort and resources.

File: packages/rocketchat-lib/server/methods/sendMessage.coffee [hunk 1]
```diff
@@ -49,8 +49,9 @@ Meteor.methods
 		message.alias = user.name if not message.alias? and RocketChat.settings.get 'Message_SetNameToAliasEnabled'
 		if Meteor.settings.public.sandstorm
 			message.sandstormSessionId = this.connection.sandstormSessionId()
-
+		
 		RocketChat.sendMessage user, message, room
+		RocketChat.metrics.messagesSent.inc()
```
Comment (graywolf336) [inline packages/rocketchat-lib/server/methods/sendMessage.coffee]: This line broke several things which rely on the return result of `RocketChat.sendMessage`.

Comment (geekgonecrazy) [inline packages/rocketchat-lib/server/methods/sendMessage.coffee]: I propose we remove the sendMessage hook here and work on a PR to make use of the statistics collection we are already creating.  That way we don't adversely effect things

**Pull request merged**